src/quickfix.c

```diff
@@ -5405,6 +5405,99 @@ get_mef_name(void)
     return name;
 }
 
+#ifdef FEAT_TERMINAL
+
+/*
+ * Run make command in a terminal buffer in a new tabpage. Returns OK on success
+ * and FAIL on failure
+ */
+    static int
+run_make_in_terminal(char_u *cmd, dict_T *dict)
+{
+    int		ret = OK;
+    jobopt_T	opt;
+    typval_T	cb_tv;
+    typval_T	dict_tv;
+    buf_T	*buf;
+#ifdef UNIX
+    char	**argv = NULL;
+    char_u	*tofree1 = NULL;
+    char_u	*tofree2 = NULL;
+
+    if (unix_build_argv(cmd, &argv, &tofree1, &tofree2) == FAIL)
+	goto fail;
+#endif
+#ifdef MSWIN
+    typval_T	argvar[2];
+
+    argvar[0].v_type = VAR_STRING;
+    argvar[0].vval.v_string = alloc(
+	    STRLEN(p_sh) + STRLEN(p_shcf) + STRLEN(cmd) + 10);
+
+    if (argvar[0].vval.v_string == NULL)
+	return FAIL;
+
+    sprintf((char *)argvar[0].vval.v_string, "%s %s %s", p_sh, p_shcf, cmd);
+    argvar[1].v_type = VAR_UNKNOWN;
+#endif
+
+    if (win_new_tabpage(0) == FAIL)
+	goto fail;
+
+    init_job_options(&opt);
+
+    // Create callback to populate location/quickfix list when make command is
+    // done.
+    cb_tv.v_type = VAR_STRING;
+    cb_tv.vval.v_string = (char_u *)"PopulateQuickFixList";
```

Comment:
Instead of splitting the terminal creation logic between the C code and the quickfix population in Vimscript, would it make sense to move the entire implementation to Vimscript and invoke the function directly?

Reply:
I see in your later comment that you are planning to move the entire logic to the C code.  That also makes sense.